File: android/app/src/main/java/org/asylumaid/net/ChatStream.kt
package org.asylumaid.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.asylumaid.AppConfig
import org.asylumaid.model.Profile
import org.asylumaid.model.StreamEvent
import org.asylumaid.model.WireMessage
import java.io.IOException
import kotlin.coroutines.coroutineContext

/** Why a chat stream ended badly. The UI maps these to localized messages. */
enum class ChatError { NETWORK, STREAM }
class ChatStreamException(val reason: ChatError) : IOException(reason.name)

/**
 * Streams `/api/chat` (newline-delimited JSON) as an ordered [Flow] of assistant text
 * deltas — the Android analogue of the web client's TextDecoder loop and iOS `bytes.lines`.
 *
 * OkHttp's streaming `ResponseBody.source()` + `readUtf8Line()` splits on `\n` and decodes
 * multi-byte UTF-8 across chunk boundaries. Each line is
 * `{type:"delta",text}` / `{type:"done"}` / `{type:"error",message}`.
 *
 * The flow throws [ChatStreamException] with [ChatError.NETWORK] if the connection fails
 * before any data, or [ChatError.STREAM] on a mid-stream / server error line.
 */
class ChatStream(private val baseUrl: String = AppConfig.BASE_URL) {

    @Serializable
    private data class RequestBody(
        val language: String,
        val profile: Profile,
        val messages: List<WireMessage>,
    )

    fun stream(
        language: String,
        profile: Profile,
        messages: List<WireMessage>,
    ): Flow<String> = flow {
        val bodyJson = appJson.encodeToString(RequestBody(language, profile, messages))
        val req = Request.Builder()
            .url(baseUrl.trimEnd('/') + "/api/chat")
            .header("Accept", "application/x-ndjson")
            .post(bodyJson.toRequestBody("application/json".toMediaType()))
            .build()

        val call = okHttp.newCall(req)
        val response = try {
            call.execute()
        } catch (e: IOException) {
            throw ChatStreamException(ChatError.NETWORK)
        }

        response.use { resp ->
            val httpOk = resp.isSuccessful
            val source = resp.body?.source() ?: throw ChatStreamException(ChatError.STREAM)
            try {
                while (true) {
                    coroutineContext.ensureActive() // cooperative cancellation
                    val line = source.readUtf8Line() ?: break
                    val trimmed = line.trim()
                    if (trimmed.isEmpty()) continue
                    val event = runCatching { appJson.decodeFromString<StreamEvent>(trimmed) }
                        .getOrNull() ?: continue // defensive: skip malformed lines
                    when (event.type) {
                        "delta" -> event.text?.let { emit(it) }
                        "done" -> return@use
                        // Server reports failures (incl. rate-limit) as an error line; the
                        // UI surfaces a localized message, not the raw server text.
                        "error" -> throw ChatStreamException(ChatError.STREAM)
                    }
                }
                // Stream ended with no explicit `done`; a non-2xx is a failure.
                if (!httpOk) throw ChatStreamException(ChatError.STREAM)
            } finally {
                call.cancel()
            }
        }
    }.flowOn(Dispatchers.IO)
}

File: ios/AsylumAid/Networking/ChatStream.swift
import Foundation

enum ChatError: Error {
    case network   // connection failed
    case stream    // server-side / mid-stream failure
}

/// Streams `/api/chat` (newline-delimited JSON) as an ordered sequence of assistant
/// text deltas.
///
/// `URLSession.bytes(for:)` + `.lines` is the NDJSON-native path: it splits on `\n`
/// and decodes multi-byte UTF-8 across chunk boundaries, exactly like the web client's
/// TextDecoder loop. Each line is `{type:"delta",text}` / `{type:"done"}` / `{type:"error",message}`.
///
/// Returning an `AsyncThrowingStream` lets the caller consume deltas on the main actor
/// in arrival order with a simple `for try await` loop.
struct ChatStream {
    var baseURL: URL = AppConfig.baseURL
    var session: URLSession = .shared

    private struct RequestBody: Encodable {
        let language: String
        let profile: Profile
        let messages: [WireMessage]
    }

    func stream(language: String,
                profile: Profile,
                messages: [WireMessage]) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var comps = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
                    comps.path = "/api/chat"

                    var req = URLRequest(url: comps.url!)
                    req.httpMethod = "POST"
                    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    req.setValue("application/x-ndjson", forHTTPHeaderField: "Accept")
                    req.httpBody = try JSONEncoder().encode(
                        RequestBody(language: language, profile: profile, messages: messages))

                    let bytes: URLSession.AsyncBytes
                    let response: URLResponse
                    do {
                        (bytes, response) = try await session.bytes(for: req)
                    } catch {
                        continuation.finish(throwing: ChatError.network); return
                    }

                    let status = (response as? HTTPURLResponse)?.statusCode ?? -1
                    let httpOK = (200..<300).contains(status)

                    for try await line in bytes.lines {
                        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                        if trimmed.isEmpty { continue }
                        guard let data = trimmed.data(using: .utf8),
                              let event = try? JSONDecoder().decode(StreamEvent.self, from: data) else {
                            continue // defensive: skip malformed lines
                        }
                        switch event.type {
                        case "delta":
                            if let text = event.text { continuation.yield(text) }
                        case "done":
                            continuation.finish(); return
                        case "error":
                            // Server reports failures (incl. rate-limit) as an error line;
                            // the UI surfaces a localized message, not the raw server text.
                            continuation.finish(throwing: ChatError.stream); return
                        default:
                            continue
                        }
                    }
                    // Stream ended with no explicit `done`; a non-2xx is a failure.
                    continuation.finish(throwing: httpOK ? nil : ChatError.stream)
                } catch is CancellationError {
                    continuation.finish(throwing: CancellationError())
                } catch {
                    continuation.finish(throwing: ChatError.stream)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}
